sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ReadChangeStreamPartitionRangeTracker.java

```diff
@@ -49,6 +54,48 @@ public ReadChangeStreamPartitionRangeTracker(PartitionMetadata partition, Timest
     this.partition = partition;
   }
 
+  /**
+   * Sets a soft timeout from now for processing new positions. After the timeout the shouldContinue
+   * will start returning false indicating an early exit from processing.
+   */
+  @Override
+  public void setSoftTimeout(Duration duration) {
+    softDeadline = new Instant(timeSupplier.get().toSqlTimestamp()).plus(duration);
+    continueProcessing = true;
+  }
+
+  /**
+   * Returns true if the restriction tracker can claim new positions.
+   *
+   * <p>If soft timeout isn't set always returns true. Otherwise:
+   *
+   * <ol>
+   *   <li>If soft deadline hasn't been reached always returns true.
+   *   <li>If soft deadline has been reached but we haven't processed any positions returns true.
+   *   <li>If soft deadline has been reached but the new position is the same as the last attempted
+   *       position returns true.
+   *   <li>If soft deadline has been reached and the new position differs from the last attempted
+   *       position returns false.
+   * </ol>
+   *
+   * @return {@code true} if the position processing should continue, {@code false} if the soft
+   *     deadline has been reached and we have fully processed the previous position.
+   */
+  @Override
+  public boolean shouldContinue(Timestamp position) {
+    if (!continueProcessing) {
+      return false;
+    }
+    if (softDeadline == null || lastAttemptedPosition == null) {
+      return true;
+    }
+
+    continueProcessing &=
+        new Instant(timeSupplier.get().toSqlTimestamp()).isBefore(softDeadline)
```

Comment:
nit: could we do the comparison without having to create a `new Instant` on every call?